ath79: add support for MikroTik SXT-R LTE

mk248 opened the pull request:
This adds support for the Mikrotik SXT-R LTE device, which is an
    outdoor router that can host an LTE modem mini PCIe card as its WAN
    connection.  The LTE antenna system claims a 9 dBi gain.

    Hardware:
        - SoC: QCA9531 @ 650MHz
        - Flash: 16 MiB SPI NOR
        - RAM: 64 MiB DDR2
        - WiFi: None.  No antenna is provided for the SoC radio.
        - Ethernet: 2 x 10/100 Mb
        - PoE: Passive or 802.3af input to port 1. Unit must be powered
	       by PoE.  GPIO switchable passive PoE out on port 2.  PoE
               output status LED blinks by hardware if overloaded.
        - Modem support: Mini PCIe slot is wired only for USB2. The PCIe
               SoC bus is not wired to the slot. Two MicroSIM slots--
               SIM card to use is selected by a GPIO output.  A GPIO can
               switch off the power to the modem.  Modem control pins
               22 and 20 (typically reset and RF kill) may be pulled to
               ground by GPIOs.
        - Serial: There is a set of 10 test pads on board near the flash
               chip.  Pin 6 is RX and pin 7 is Tx.  Tx GPIO pin (10) is
               shared with the yellow LED on the eth1 socket.  This DTS
               sets it as serial output allowing OpenWrt to use the
               port.  The bootloader does not use serial.
        - LEDs: Power, LTE, 2x Ethernet Link, PoE out, 5x RSSI bars
        - Buttons: Reset, Mode

    Installation:  Standard MikroTik-- RouterBOOT netboot the OpenWrt
        initramfs image, then flash the sysupgrade image from within
        OpenWrt.  Under OpenWrt, the eth1 port (with PoE in) defaults to
        be LAN, and eth2 is WAN.

    Signed-off-by: Mike Keitz <keitz24@gmail.com>

mans0n commented:
Hi, this has some formal issues, especially:
* Please fix the commit title.
* Please fix indentation (space -> tab for board.d/*) and remove trailing spaces in DTS.

mk248 commented:
Sorry, I had to step away from this for a bit as it is very frustrating.  I thought that git commit --amend could be used to revise the commit, that worked only once though and now I find that Geany plays havoc with tabs vs. spaces so I need to do it again.

What is the proper way to update the pull request?  Would I re-checkout my feature branch then re-merge it?

mans0n commented:
You should drop the merge commit. Rebase your commit on top of the current master and do force push.

While at it, could you also de-indent the commit message? (first 4 spaces in each line)

mk248 committed:
ath79: add support for MikroTik SXT-R LTE

This adds support for the MikroTik SXT-R LTE device, which is an
outdoor router that can host an LTE modem mini PCIe card as its WAN
connection.  The LTE antenna system claims a 9 dBi gain.

Hardware:
- SoC: QCA9531 @ 650MHz
- Flash: 16 MiB SPI NOR
- RAM: 64 MiB DDR2
- WiFi: None.  No RF/antenna circuitry on the SoC radio pins.
- Ethernet: 2 x 10/100 Mb
- PoE: Passive or 802.3af input to port 1. Unit must be powered
    by PoE.  GPIO switchable passive PoE out on port 2.  PoE
    output status LED blinks by hardware if overloaded.
- Modem support: Mini PCIe slot is wired only for USB2. The PCIe
    SoC bus is not wired to the slot. Two MicroSIM slots--
    SIM card to use is selected by a GPIO output.  A GPIO can
    switch off the power to the modem.  Modem control pins
    22 and 20 (typically reset and RF kill) may be pulled to
    ground by GPIOs.
- Serial: There is a set of 10 test pads on board near the flash
    chip.  Pin 6 is RX and pin 7 is Tx.  Tx GPIO pin (10) is
    shared with the yellow LED on the eth1 socket.  This DTS
    sets it as serial output allowing OpenWrt to use the
    port.  The bootloader does not use serial.
- LEDs: Power, LTE, 2x Ethernet Link, PoE out, 5x RSSI bars
- Buttons: Reset, Mode

Installation:  Standard MikroTik-- RouterBOOT netboot the OpenWrt
   initramfs image, then flash the sysupgrade image from within
   OpenWrt.  Under OpenWrt, the eth1 port (with PoE in) defaults to
   be LAN, and eth2 is WAN.

Signed-off-by: Mike Keitz <keitz24@gmail.com>

mk248 commented:
It seems to have closed when I deleted the first merge.  Now trying to reopen.

f00b4r0 commented on target/linux/ath79/dts/qca9531_mikrotik_routerboard-sxtr.dts:
```diff
@@ -0,0 +1,189 @@
+
+&eth0 {
+    status = "okay";
+    mtd-mac-address = <&hard_config 0x10>;
+    mtd-mac-address-increment = <1>;
```
Don't do this. See how it's done for other mikrotik devices using `02_network`.
By the way `mtd-mac-address` has been removed from the tree (in 5ae2e786), so this suggests you didn't test your changes.

f00b4r0 commented on target/linux/ath79/mikrotik/base-files/etc/board.d/02_network:
```diff
@@ -23,6 +23,9 @@ ath79_setup_interfaces()
 		ucidef_set_interface_lan "eth0"
 		;;
+	mikrotik,routerboard-sxtr)
+		ucidef_set_interfaces_lan_wan "eth1" "eth0"
```
Given the description of the device, shouldn't both interfaces be attached to LAN by default?

f00b4r0 commented on target/linux/ath79/dts/qca9531_mikrotik_routerboard-sxtr.dts:
```diff
@@ -0,0 +1,189 @@
+        };
+
+        eth1green {
+            label = "green:eth1";
```
This naming might be confusing with linux interface name.
Aren't the ports labelled? The product webpage mentions 'ether1' and 'ether2', might be more appropriate if it matches device labelling. This applies to LED netdev triggers as well.

f00b4r0 reviewed (changes requested):
Thanks for your submission.
There are a few problems, see comments.

mans0n commented on target/linux/ath79/dts/qca9531_mikrotik_routerboard-sxtr.dts:
```diff
@@ -0,0 +1,189 @@
+    #size-cells = <0>;
+    status = "okay";
+
+    hub_port: port@1 {
+        eg = <1>;
+        #trigger-source-cells = <0>;
+    };
```
`reg`?

It seems this node is not referenced anywhere, so you can simply drop it.
```diff
&usb0 {
	status = "okay";
};
```

mans0n commented:
Please use tabs instead of spaces for indents in DTS.

I'm not familiar with mikrotik devices so I can't do much review here, sorry.

mk248 commented:
Thank you both for your comments  I will get back into the code and make the requested changes within a few days.

* I did check that the MAC somehow got assigned as the sticker MAC but I will do it in the board file instead.  I do still have the device available for testing.
* The physical marking on the ports is exactly "ETH 1" and "ETH 2". 
* I think hub_port has something to do with USB.  It is necessary to enable USB in this dts file since the dtsi files for other Mikrotik QCA9531 models leave it disabled, and the modem will not work without it.  I copied the USB section from the dts for a TP-Link using the same SoC.  I will test without hub_port and confirm if it still works.
* In the final use case the user would likely want both Ethernet ports bridged into lan.  But since no LTE modem drivers are in the distribution package, the user would want a non LTE way to put the router online to install drivers for their modem.  Having the other port configured as a wired WAN will facilitate that.  Also this is how most models with two or more Ethernet ports default.

f00b4r0 commented:
> * I did check that the MAC somehow got assigned as the sticker MAC but I will do it in the board file instead.  I do still have the device available for testing.

It may work for your particular device, but there is no guarantee that the MAC address will be stored at the exact same offset in a different device. In fact it likely won't. That's why we set the MAC address via sysfs + board file on Mikrotik devices.

> * The physical marking on the ports is exactly "ETH 1" and "ETH 2".

ok that's slightly unfortunate. I don't know what the general consensus is but I would avoid using LED names that look like linux network device names. Maybe spell them out partly or in full "ether1" or "ethernet1"? Or use an underscore "eth_1"? Anything that doesn't look like `eth1` is preferable IMHO.

> * In the final use case the user would likely want both Ethernet ports bridged into lan.  But since no LTE modem drivers are in the distribution package, the user would want a non LTE way to put the router online to install drivers for their modem.  Having the other port configured as a wired WAN will facilitate that. 

I don't see why. There is no difficulty in downloading anything over LAN, in fact single-port APs are perfectly capable of downloading software provided their network is correctly setup. In any case a hypothetical use case should not be the driving factor in this decision.

The device markings do not actively discern LAN and WAN and therefore it seems logical that both ports be in LAN by default; that would be my recommendation.

Otherwise the next question that will likely come is "why do you chose ether1 (PoE In) as LAN and ether2 as WAN and not the opposite?", especially when other MikroTik devices with PoE In and distinctly named WAN ports have the WAN port as PoE in (case in point the hAP, the hEX...).

> Also this is how most models with two or more Ethernet ports default.

When it makes sense (typically because the ports are so named). There are also valid examples of 2-port devices with both ports in LAN, see:
https://github.com/openwrt/openwrt/blob/0d2d52df69e91fe2f6e3e65750c0728fc45f5ab2/target/linux/ath79/generic/base-files/etc/board.d/02_network#L124-L132

Incidentally, most of them are outdoor APs like the SXT-R.

HTH

f00b4r0 commented:
> > * The physical marking on the ports is exactly "ETH 1" and "ETH 2".
> 
> ok that's slightly unfortunate. I don't know what the general consensus is but I would avoid using LED names that look like linux network device names. Maybe spell them out partly or in full "ether1" or "ethernet1"? Or use an underscore "eth_1"? Anything that doesn't look like `eth1` is preferable IMHO.

FWIW I'm noticing that with some other Mikrotik devices in ar71xx we used "port1", "port2" etc.
Seems like a valid choice too.
